### PR TITLE
feat(enhancement): Load plugin metadata (name, about) from plugin.txt.

### PR DESCRIPTION
**Pull request by Beed-git**
-----------------------
**Feature:** 

## Feature Details
This is a proposal to add an optional (but recommended) plugin.txt file to the root folder of each plugin. This file would take over the duties of the about.txt file and allows the plugin authors to override the name of their plugin. This allows the plugins to have a different name to the name of their folder, and allow for plugins to share the same name. This system could be expanded in the future to have other information about the author (such as their name, a link to their website, Discord, etc.) or hold a list of required, dependent plugins used by the plugin.

First part of a plugin overhaul as described in: https://github.com/endless-sky/endless-sky/discussions/8932

## Usage Examples

A theoretical plugin manager downloads mod #123456 from its server, the root folder for the plugin is set to this id to avoid conflicts with plugins of similar names. Opening the plugin in the plugins window shows the name and about text is set to the metadata within the plugin.txt file.

image

## Testing Done
Added 3 plugins to the plugins folder, one which correctly uses the new plugin format, one which uses a broken plugin.txt with missing and broken fields, and one which doesn't have a plugin.txt.

When opening the plugin UI in preferences, we see that first plugin has the correct name and about which is set from its plugin.txt, while the other two fall back to using their folder names as their plugin names, and loads their about information from about.txt.  

Plugin A: 
image

Plugin B:
image

Plugin C:
image

Plugins folder:
image

(An empty plugin.txt has the same result as Plugin C.)

## Performance Impact
There may be a slight increase in loading times for users on old/slow hardware with hundreds of plugins as we load a file for every plugin but the impact should be negligible.

### CODE REVIEW

**Comment by Zitchas** [comment]
This sounds interesting. I'm curious if there's any potential overlap with the plugin index data or way we could use the two of them together.

**Comment by Beed-git** [comment]
> This sounds interesting. I'm curious if there's any potential overlap with the plugin index data or way we could use the two of them together.

One solution could be to move most of the information within the manifest files into the plugin.txt, that way the manifest would only need to contain a link to the repo. Or alternatively the plugin.txt could be generated by the manifest file. 

Either way there are definitely plenty of ways the two can be integrated.

**Comment by quyykk** [comment]
This is definitely interesting! I'll tag @MCOfficer too for his thoughts.

**Comment by lumbar527** [comment]
Wait, does this mean `about.txt` would not be used anymore? That's unclear to me.

**Comment by Beed-git** [comment]
> Wait, does this mean `about.txt` would not be used anymore? That's unclear to me.

Yep, so the idea is that instead of having the about.txt you would instead write the about text in the plugin.txt.  (For backwards compatibility reasons about.txt does still work, but the about field in plugin.txt takes priority.)

**Comment by MCOfficer** [comment]
Conceptually, this is sound, unless we collectively decide to have a more advanced, more machine-readable file in its place. But since i'm against that, this lgtm.

One thought: please add some debug logging in case of a broken plugin.txt, so the implicit fallback to folder names is easier to resolve for plugin authors.

File: source/Plugins.cpp [hunk 1]
```diff
@@ -65,10 +66,36 @@ const Plugin *Plugins::Load(const string &path)
...
 
-	auto *plugin = plugins.Get(name);
-	plugin->name = std::move(name);
+	auto * plugin = plugins.Get(name);
+
+	string pluginFile = path + "/plugin.txt";
+	bool hasPluginFile = Files::Exists(pluginFile);
+
+	// Loads plugin metadata from plugin.txt.
+	DataFile file(path + "/plugin.txt");
```
Comment (quyykk) [inline source/Plugins.cpp:75]:
```suggestion
	DataFile file(pluginFile);
```

File: source/Plugins.cpp [hunk 2]
```diff
@@ -65,10 +66,36 @@ const Plugin *Plugins::Load(const string &path)
 	size_t pos = path.rfind('/', path.length() - 2) + 1;
 	string name = path.substr(pos, path.length() - 1 - pos);
 
-	auto *plugin = plugins.Get(name);
-	plugin->name = std::move(name);
+	auto * plugin = plugins.Get(name);
+
+	string pluginFile = path + "/plugin.txt";
+	bool hasPluginFile = Files::Exists(pluginFile);
```
Comment (quyykk) [inline source/Plugins.cpp:72]: Since this is just used once, it doesn't have to be a variable.

File: source/Plugins.cpp [hunk 3]
```diff
@@ -65,10 +66,36 @@ const Plugin *Plugins::Load(const string &path)
...
+
+	// Loads plugin metadata from plugin.txt.
+	DataFile file(path + "/plugin.txt");
+	for(const DataNode & child : file)
```
Comment (quyykk) [inline source/Plugins.cpp:76]:
```suggestion
	for(const DataNode &child : file)
```

Comment (RisingLeaf) [reply source/Plugins.cpp:76]: Normally the first layer would be called node, right?
Progression like this: node, child, grand

Comment (quyykk) [reply source/Plugins.cpp:76]: yup

File: source/Plugins.cpp [hunk 4]
```diff
@@ -66,9 +67,34 @@ const Plugin *Plugins::Load(const string &path)
...
+	if(plugin->name.empty())
+	{
+		plugin->name = std::move(name);
+		if(Files::Exists(pluginFile))
+		{
+			Logger::LogError(
+				"Failed to find name field in plugin.txt. Defaulting plugin name to folder name: \"" + plugin->name + "\"");
+		}
+	}
+	if(plugin->aboutText.empty())
```
Comment (Amazinite) [inline source/Plugins.cpp:94]:
```suggestion
	// Load plugin metadata from plugin.txt.
	DataFile file(pluginFile);
	for(const DataNode &child : file)
	{
		if(child.Token(0) == "name" && child.Size() >= 2)
			plugin->name = child.Token(1);
		else if(child.Token(0) == "about" && child.Size() >= 2)
			plugin->aboutText = child.Token(1);
	}

	// Set missing required values.
	if(plugin->name.empty())
	{
		plugin->name = std::move(name);
		if(Files::Exists(pluginFile))
		{
			Logger::LogError(
				"Failed to find name field in plugin.txt. Defaulting plugin name to folder name: \"" + plugin->name + "\"");
		}
	}
	// Set values from old about.txt files.
	if(plugin->aboutText.empty())
```

**Comment by Pshy0** [comment]
How do we make a multi-line description with this new system?
In the about.txt file, it was possible to use line breaks (`\n`).

**Comment by lumbar527** [comment]
Using line breaks doesn't seem to work.

**Comment by Amazinite** [comment]
Yeah, line breaks won't work with this. The `about` node needs to be updated to behave like the `description` node of outfits and other objects in game, where a new node will append to whatever came before it.
https://github.com/endless-sky/endless-sky/blob/cd8b71c95325d626030b9b067245d729186b8d1a/source/Outfit.cpp#L266-L270

**Pull request merged**